feat(ow-deletions): Add Overwatch notification on organization deletion

tnt-sentry opened the pull request:
## Summary

Implements webhook notifications to Overwatch when organizations are deleted in Sentry. This builds on top of the Overwatch webhook forwarding infrastructure from #99389, enabling Overwatch to clean up organization-specific data when an organization is removed from Sentry.

### Core Implementation

- **Celery Task** (`src/sentry/deletions/tasks/overwatch.py`): New async task that sends organization deletion notifications to configured Overwatch regions
- **Deletion Hook** (`src/sentry/deletions/defaults/organization.py`): Modified `OrganizationDeletionTask.delete_instance()` to trigger webhook notifications via `transaction.on_commit()`

tnt-sentry committed:
feat(deletions): Add Overwatch notification on organization deletion

Add webhook notification to Overwatch when organizations are deleted.
Each region sends a notification to its local Overwatch instance with
organization details.

- Implement async Celery task with HMAC-SHA256 signature authentication
- Hook into OrganizationDeletionTask.delete_instance() using transaction.on_commit
- Include comprehensive unit and integration tests

tnt-sentry committed:
fix(overwatch): Add error handling for region retrieval in organization deletion notifications

Enhance the notify_overwatch_organization_deleted function to include error handling when retrieving the local region. Log exceptions with relevant organization details to improve debugging and monitoring of Overwatch notifications.

tnt-sentry committed:
feat(deletions): Add Overwatch notification on organization deletion

Add webhook notification to Overwatch when organizations are deleted.
Each region sends a notification to its local Overwatch instance with
organization details.

- Implement async Celery task with HMAC-SHA256 signature authentication
- Hook into OrganizationDeletionTask.delete_instance() using transaction.on_commit
- Include comprehensive unit and integration tests
- Use OVERWATCH_REGION_URL setting for the Overwatch endpoint

ajay-sentry commented on src/sentry/deletions/tasks/overwatch.py:
```diff
@@ -0,0 +1,142 @@
+            },
+        )
+        metrics.incr(
+            "overwatch.organization_deleted.sent",
```
should this be deleted.failed?

ajay-sentry commented on src/sentry/deletions/tasks/overwatch.py:
```diff
@@ -0,0 +1,142 @@
+            },
+        )
+        metrics.incr(
+            "overwatch.organization_deleted.sent",
```
do we care about separating these metric failures from the other?

ajay-sentry commented on tests/sentry/deletions/tasks/test_overwatch.py:
```diff
@@ -0,0 +1,312 @@
+        with self.options({"overwatch.enabled-regions": ["us"]}):
+            notify_overwatch_organization_deleted(self.organization_id, self.organization_slug)
+
+        # Verify the request was made
```
can probably nix a lot of these comments

ajay-sentry commented on tests/sentry/deletions/tasks/test_overwatch.py:
```diff
@@ -0,0 +1,312 @@
+    def test_successful_notification(
+        self, mock_metrics: Any, mock_logger: Any, mock_get_local_region: Any
+    ) -> None:
+        """Test successful notification to the local region"""
```
also not a fan of the test doc string personally, but open to leaving those if you prefer

tnt-sentry committed:
fix: Ensure implementation matches test expectations

- Move OVERWATCH_REGION_URL check after region check to match test order
- Add debug logging when region is not enabled
- Use 'missing_region_url' log message to match tests
- Add logger mock to test_different_region_name test
- Remove sample_rate parameter from metrics call

ajay-sentry reviewed (approved):
looks good

tnt-sentry committed:
fix: Use captureOnCommitCallbacks in organization deletion test

The test was failing because transaction.on_commit callbacks don't
execute automatically in tests. Use captureOnCommitCallbacks context
manager to ensure the Overwatch notification task is called during
the test.

Also capture org_id before deletion since the org instance won't
exist after the deletion completes.

tnt-sentry committed:
fix: Patch transaction.on_commit to execute immediately in test

Instead of using captureOnCommitCallbacks which may not be available,
directly patch transaction.on_commit in the organization module to
execute callbacks immediately during the test. This ensures the
Overwatch notification task is called when the organization is deleted.

tnt-sentry committed:
fix: Test delete_instance directly to avoid transaction issues

Instead of testing the full deletion flow with transaction.on_commit
complications, test the OrganizationDeletionTask.delete_instance method
directly. This allows us to:
- Patch transaction.on_commit in a controlled scope
- Execute the callback immediately for testing
- Avoid interference with other transaction callbacks (like drain_shard)
- Remove unused imports (responses, typing.Any)

tnt-sentry committed:
fix: Add type annotations to test callback function

Add typing.Any import and type annotations to the execute_callback
function to satisfy mypy type checking requirements.
